In [47]:
# Imports
import pandas as pd
from sqlalchemy import create_engine, inspect


In [48]:
# Create connection and read data
engine = create_engine('postgresql://neondb_owner:npg_CeS9fJg2azZD@ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech:5432/neondb?sslmode=require')

# Inspect schema for table names
inspector = inspect(engine)

# List schemas (optional)
schemas = inspector.get_schema_names()
print("Schemas:", schemas)

# List tables in a specific schema (e.g., "nyc_schools")
tables = inspector.get_table_names(schema="nyc_schools")
print("Tables in 'nyc_schools':", tables)


Schemas: ['dependency_example', 'information_schema', 'nyc_schools', 'public', 'test_berlin_data']
Tables in 'nyc_schools': ['school_safety_report', 'high_school_directory', 'school_demographics']


In [49]:
# Query data
df = pd.read_sql('SELECT * FROM nyc_schools.high_school_directory LIMIT 5', engine)
print(df.head(5))

      dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

  expgrade_span_min expgrade_span_max  ... number_programs  \
0           

In [50]:
# 1.How many schools are there in each borough?

query = '''
SELECT borough, COUNT (DISTINCT dbn) AS school_count
FROM nyc_schools.high_school_directory
GROUP BY borough
ORDER BY school_count DESC
'''

df = pd.read_sql(query, engine)
print(df)

         borough  school_count
0       Brooklyn           121
1          Bronx           118
2      Manhattan           106
3         Queens            80
4  Staten Island            10


In [59]:
# 2.What is the average % of English Language Learners (ELL) per borough?

query = '''
SELECT h.borough, AVG(d.ell_percent) AS avg_ell_percent
FROM nyc_schools.high_school_directory h
FULL JOIN nyc_schools.school_demographics d ON h.dbn = d.dbn
WHERE d.ell_percent IS NOT NULL
GROUP BY h.borough
ORDER BY h.borough;
'''


df = pd.read_sql(query, engine)
display(df)

borough  avg_ell_percent
0  Manhattan         7.572500
1       None        12.708867

In [65]:
# 3.Using the data from the school demographics and high school directory, write a query to find the top 3 schools in each borough with the highest percentage
# of special education students (sped_percent)

query = '''
SELECT d.schoolyear, d.dbn, h.borough, d."Name",  d.sped_percent,
DENSE_RANK() OVER (PARTITION BY h.borough ORDER BY d.sped_percent DESC) AS sped_rank
FROM nyc_schools.school_demographics AS d
JOIN nyc_schools.high_school_directory AS h
ON d.dbn = h.dbn
WHERE d.sped_percent IS NOT NULL
ORDER BY sped_rank
LIMIT 3
'''

df = pd.read_sql(query, engine)
display(df)

schoolyear     dbn    borough                             Name  \
0    20092010  01M450  Manhattan  EAST SIDE COMMUNITY HIGH SCHOOL   
1    20102011  01M450  Manhattan  EAST SIDE COMMUNITY HIGH SCHOOL   
2    20082009  01M450  Manhattan  EAST SIDE COMMUNITY HIGH SCHOOL   

   sped_percent  sped_rank  
0          28.8          1  
1          27.7          2  
2          26.7          3

### **1. Number of Schools per Borough**

| Borough       | Number of Schools |
| ------------- | ----------------- |
| Brooklyn      | 121               |
| Bronx         | 118               |
| Manhattan     | 106               |
| Queens        | 80                |
| Staten Island | 10                |

Insight:
Brooklyn has the highest number of schools (121), while Staten Island has the fewest (10). The distribution aligns with population density and urban development.

---

### **2. Average % of English Language Learners (ELL) per Borough**

| Borough   | Avg. ELL % |
| --------- | ---------- |
| Manhattan | 7.57%      |
| *None*    | 12.71%     |

Insight:
Only Manhattan shows a valid borough-wide ELL average (\~7.6%). The presence of `None` suggests missing or unmatched borough data — possibly due to incomplete records.

---

### **3. Top 3 Schools by % of Special Education Students**

| Year      | DBN    | School Name                     | SpEd % | Rank |
| --------- | ------ | ------------------------------- | ------ | ---- |
| 2009–2010 | 01M450 | East Side Community High School | 28.8%  | 1    |
| 2010–2011 | 01M450 | East Side Community High School | 27.7%  | 2    |
| 2008–2009 | 01M450 | East Side Community High School | 26.7%  | 3    |

Insight:
All top 3 records for Manhattan are from the same school (`01M450 – East Side Community HS`), indicating consistently high support for special education students over multiple years.
